packages/block-library/src/image/save.js

```diff
@@ -73,5 +74,12 @@ export default function save( { attributes } ) {
 		);
 	}
 
+	if ( isListItem ) {
+		return (
+			<li className={ classes }>
+				<figure>{ figure }</figure>
+			</li>
```

Comment:
On first glace, I would think that, similar to how the "block alignment" wrapping `div` works (see the code a few lines earlier), the classes should still be applied to the `figure`, rather than the `li`.

Additionally, I wonder what consequences this may have for styles targeting the `wp-block-image` class, since that could now apply to an `li` in some cases, rather than the `figure`. Either way, I suppose theme style issues are inevitable with this sort of massive change.

Looking at the code, it looks like alignment wrappers are [specially handled](https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/block-list/block.js#L189) by the `BlockListBlock` component to ensure the editor-only `wp-block` CSS class is applied to the alignment wrapper and not the wrapped contents.

Replicating the same behavior in this case would probably require a similar modification. But modifying that general component just for this one use-case would obviously be a hack.

However, if we ended up taking the hypothetical `<InnerBlocks childWrapper="li" />` kind of approach instead of using block context, the block classes could stay on the `figure` with no problems.

Another point in favor of the `InnerBlocks`-prop approach: even if we generalize the block context name for this list-item wrapping, it's still left up to the individual blocks to implement the `li` wrapper, and it opens the door to inconsistent handling of where the block classes are put.

Additionally, the `InnerBlocks`-prop approach would allow this child-wrapper behavior to be generalized even further to cases where the wrapper could even have multiple levels, e.g. `childWrapper={ MyWrapper }` where `MyWrapper` is a component that returns something like `<li><div>{ theChildBlock }</div></li>`.

Another benefit would be that the child blocks would not need to be aware of the parent block to take advantage of this. The parent block would This would allow the Gallery block to more easily support 3rd-party image blocks without each of those blocks having to handle the `li` wrapping themselves.

Yet another benefit: with block context, something like `isList` would be passed down to not only direct children, but also children-of-children. Since the direct children may not themselves be lists, those children would have to override `isList` themselves to prevent their children from inheriting the same value. Unlike something like `postId` or `postType`, the child block wrapper is only relevant to the parent and its direct children. The `InnerBlocks`-prop approach doesn't have this problem.

Ultimately, the list-element wrapper feels more like a part of the parent, rather than the child. Though implementing an `InnerBlocks` `childWrapper` prop may result in a slight increase to the complexity of `InnerBlocks`, it would solve the issue in one spot and prevent other blocks from having to reimplement it over and over again, and it would remove the possibility for inconsistent handling of where the block classes get put.

I wasn't really sure before, but now I'm convinced that the `InnerBlocks` approach is the way to go.

Reply:
yeh, an `InnerBlocks` `childwrapper` prop would be a nice tidy decoupled way to handle this ... do you have any plans to explore adding this to InnerBlocks?

Reply:
It's completely untested and doesn't work for React Native yet, but I went ahead and threw together #24232 to implement that very feature.

Try rebasing your branch off of mine. You should be able to use the prop like so:

```jsx
<InnerBlocks __experimentalItemWrapper="li" />
```
```jsx
<InnerBlocks.Content __experimentalItemWrapper="li" />
```

Let me know if it works.

Reply:
sorry, got sidetracked with other project work today, will try and give it a go tomorrow

Reply:
That's okay; my branch had a bunch of obvious bugs until recently anyway. :stuck_out_tongue:

Reply:
@ZebulanStanphill  - I gave it a try combined with your PR and it seemed to work well for me. I will wait a few more days to see how the conversation proceeds on your PR before I make any permanent changes to this PR.

Reply:
@ZebulanStanphill I have updated this to use your new _experimentalItemCallback prop. It seems to work, so will try and get some more tidying up done on this PR and get it to a point were we can get it properly reviewed.

Reply:
@ZebulanStanphill The _experimentalItemCallback prop doesn't seem to work on InnerBlocks.Content, I added it on this PR to the gallery save method, but the saved block content, and front end render doesn't include the <li> wrapper ... I am probably misunderstanding or missing something, so if you can take a look when you get a chance that would be great thanks

Reply:
I just tried out the branch myself, and the `li` wrapping on the front-end seems to be working fine.

Reply:
hmm, working for me today ... go figure 🤷

Reply:
@ZebulanStanphill for me `__experimentalItemCallback` in `InnerBlocks.Content` in the save function works fine when adding the block to to a page then viewing in the frontend, but it causes the block validation to fail when reloading a page with the gallery block in the editor:

```
Content generated by `save` function:

<div class="wp-block-gallery columns-3 alignnone is-cropped is-style-tiled"></div>

Content retrieved from post body:

<figure class="wp-block-gallery columns-3 is-cropped is-style-tiled"><ul><li class="wp-block-gallery__image"></li>

<li class="wp-block-gallery__image"></li>

<li class="wp-block-gallery__image"></li></ul></figure>
```

But the funny thing is that if you select the 'Resolve` option bother the before and after look the same:

<img width="813" alt="Screen Shot 2020-08-21 at 3 18 37 PM" src="https://user-images.githubusercontent.com/3629020/90848751-bdc1a800-e3c1-11ea-8d79-b61df3eec4df.png">

Will take another look at it next week, but if you have some clues as to why this is happening that would be great 😄

Reply:
>Did 801a6cd fix the issue?

No it didn't unfortunately ... haven't had chance to get back and have a more detailed look at it yet.

Reply:
I just tested it out. When I refresh the editor, it just fails to load, with this JS error occurring:
![image](https://user-images.githubusercontent.com/19592990/91232752-58d5bd00-e6f5-11ea-87d5-e6c6c318b810.png)

It looks like the problem is happening at [this line](https://github.com/WordPress/gutenberg/pull/24039/files#diff-cf0a8b79f37693f7ad2f99b386dd470cR7), which gets called as a result of [this line](https://github.com/WordPress/gutenberg/pull/24039/files#diff-e25f8ba6320e6c1650e6ec1c0ae5ba34R32).

Reply:
Fixed this issue, but still getting the failing block validation ... will try and take a closer look at it this week.

Reply:
Got as far as working out that it is something to do with adding `__experimentalItemCallback` to `InnerBlocks.Content` in the `save` method. 

Without this the block saves and validates fine on editor refresh, but of course then the nested image blocks our not wrapped in `li` in the frontend.

With the callback added the nested image blocks are correctly wrapped in `<li>`s on the front end, but block validation fails when reloading the editor.

Reply:
I looked into this, and I think the problem is not with your PR, but with the `__experimentalItemCallback` API; namely, I forgot to change any of the parsing code in #24232, so the editor doesn't know how to parse the post's comment-delimited HTML properly when the render callback API is used. See [this comment](https://github.com/WordPress/gutenberg/pull/24232#issuecomment-688843373). 😟

Reply:
whew, not just something simple/stupid that I was overlooking at least 😄